### DISCUSSION TITLE
”Clients” - Collection of common Dependency ”Clients” shared and maintained by PointFreeCo and TCA community?

### DISCUSSION THREAD

**Discussion by Sajjon** [pointfreeco/swift-composable-architecture#1662] [Ideas]
A lot of us TCA users implement our own version of "Clients" e.g. UserDefaultClient, KeychainClient, PasteboardClient etc. I usually just copy paste what PointFreeCo has written in Isowords or in examples and make appropriate adjustments. 

TCA already packages some dependencies for us to use, e.g. `mainQueue`.

Would it not be useful to have a large collection of common clients as a repo? With loads of unit tests so that all of us TCA users do not reinvent the wheel ourselves in every project?

Which clients would you like to see in this repo?

**Comment by Nevwild** [comment]
This is a great idea! Luckily for us, several clients have already been open sourced [The Composable Architecture Kit](https://tcakit.com/), and you can find the associated git repo [here](https://github.com/joeblau/swift-composable-kit).  I have used some of the clients and they are great! A few have been good starting points for building out more robust clients. I think that they are all Combine based, but if we wanted to upgrade them to Swift Concurrency as a community this could be a great starting point.

**Reply by Sajjon** [reply]
I Was thinking primarily structured concurrency clients :)

**Comment by mbrandonw** [comment]
This does sound like a good idea. I think it could really be handy once we release the dependencies library as a standalone package so that people don't even have to depend on TCA. We plan on doing that very soon.

**Comment by Sajjon** [comment]
Other _potentially_ useful clients:
HapticFeedbackClient
LocalAuthenticationClient
FileImporterClient
FileExporterClient
InputValidationClient (email, phone, credit card etc)?

**Comment by saroar** [comment]
My [CommonTCALibraries](https://github.com/AddaMeSPB/CommonTCALibraries) for every one 
Lists
<img width="270" alt="Screenshot 2022-12-09 at 08 11 44" src="https://user-images.githubusercontent.com/8770772/206628780-96667934-2626-4e65-94d5-ac06479c7dbf.png">

**Comment by davdroman** [comment]
@tgrapperon just released [swift-dependencies-additions](https://github.com/tgrapperon/swift-dependencies-additions) to go along with the newly standalone [swift-dependencies](https://github.com/pointfreeco/swift-dependencies) library 🎉

**Reply by acosmicflamingo** [reply]
tgrapperon contributes so much that I am starting to recognize his username just as quickly as stephencelis and mbrandonw in the email notifications LOL

**Reply by tgrapperon** [reply]
Thanks for the shoutout @davdroman and @acosmicflamingo!
Everyone can feel free to discuss individual dependencies in the discussions there!
The objective is to maintain a relatively coherent and standardized API overall, so it could help lowering the global learning curve.

**Reply by Sajjon** [reply]
A really good Keychain client would be a golden addition!

**Reply by jshier** [reply]
There are many wrappers out there already. You simply need to define your interface. There really doesn't need to be wrappers for everything here.
